File: benchmarks/rust-bench/src/tests/ping_latency.rs
use crate::config::Cli;
use crate::protocol::{self, WsStream};
use crate::report::{LatencySummary, TierResult};
use crate::stats::{self, LatencyHistogram};
use futures_util::{SinkExt, StreamExt};
use tokio_tungstenite::tungstenite::protocol::Message;

const PINGS_PER_CONNECTION: usize = 50;

/// Test 2: Ping/Pong Latency Under Load
/// N persistent connections, each pinging multiple times. Measures tail latency.
pub async fn run(cli: &Cli) -> Vec<TierResult> {
    stats::print_header(
        "TEST 2: Ping/Pong Latency Under Concurrent Load",
        &format!(
            "N persistent connections, each pinging {}x. Measures tail latency.",
            PINGS_PER_CONNECTION
        ),
    );

    let tiers = cli.tiers_for(crate::config::TestName::PingLatency);
    let token = crate::jwt::generate_bench_token(cli.secret.as_bytes(), "bench-user");
    let mut results = Vec::new();

    for &n in &tiers {
        println!("\n  --- {} concurrent connections ---", n);

        let connections = protocol::connect_batch(
            &cli.host,
            cli.port,
            &token,
            n,
            cli.batch_size,
            "compression=false&protocol_version=1",
        )
        .await;

        if connections.is_empty() {
            println!("    FAILED: no connections");
            continue;
        }

        let actual = connections.len();

        // Each connection pings PINGS_PER_CONNECTION times
        let mut handles = Vec::with_capacity(actual);
        for ws in connections {
            handles.push(tokio::spawn(ping_loop(ws)));
        }

        let task_results = futures_util::future::join_all(handles).await;

        // Merge all histograms
        let mut merged = LatencyHistogram::new();
        let mut ws_list = Vec::new();
        for (hist, ws) in task_results.into_iter().flatten() {
            merged.merge(&hist);
            ws_list.push(ws);
        }

        println!("    Total pings: {}", merged.len());
        merged.print_summary("    ");

        results.push(TierResult {
            tier: n,
            connected: actual,
            errors: n - actual,
            duration_secs: 0.0,
            messages_sent: merged.len(),
            messages_received: merged.len(),
            bytes_sent: 0,
            latency: Some(LatencySummary::from_histogram(&merged)),
            extra: serde_json::json!({}),
        });

        protocol::close_all(ws_list).await;
        println!();
    }

    results
}

async fn ping_loop(mut ws: WsStream) -> (LatencyHistogram, WsStream) {
    let mut hist = LatencyHistogram::new();

    for _ in 0..PINGS_PER_CONNECTION {
        let now_ms = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        let ping_msg = serde_json::json!({
            "t": "ping",
            "p": { "client_timestamp": now_ms }
        });

        if ws
            .send(Message::Text(ping_msg.to_string().into()))
            .await
            .is_err()
        {
            break;
        }

        // Wait for PONG (with timeout)
        let deadline = tokio::time::Instant::now() + std::time::Duration::from_secs(5);
        while let Ok(Some(Ok(msg))) = tokio::time::timeout_at(deadline, ws.next()).await {
            if let Some(parsed) = protocol::parse_wse_message(&msg) {
                if protocol::extract_pong_timestamp(&parsed).is_some() {
                    let actual_rtt = std::time::SystemTime::now()
                        .duration_since(std::time::UNIX_EPOCH)
                        .unwrap()
                        .as_millis() as u64;
                    let rtt = actual_rtt.saturating_sub(now_ms) as f64;
                    hist.record_ms(rtt);
                    break;
                }
            }
            // Not a pong, keep reading
        }
    }

    (hist, ws)
}

File: benchmarks/ts-bench/src/tests/ping-latency.ts
import type { Config } from '../config.js';
import type { TierResult } from '../report.js';
import * as protocol from '../protocol.js';
import { LatencyHistogram, printHeader } from '../stats.js';
import { generateBenchToken } from '../jwt.js';
import { tiersFor } from '../config.js';
import type WebSocket from 'ws';

const PINGS_PER_CONNECTION = 50;

export async function run(config: Config): Promise<TierResult[]> {
  printHeader(
    'TEST 2: Ping/Pong Latency Under Concurrent Load',
    `N persistent connections, each pinging ${PINGS_PER_CONNECTION}x. Measures tail latency.`
  );

  const tiers = tiersFor(config, 'ping-latency');
  const token = generateBenchToken(Buffer.from(config.secret), 'bench-user');
  const results: TierResult[] = [];

  for (const n of tiers) {
    console.log(`\n  --- ${n} concurrent connections ---`);

    const connections = await protocol.connectBatch(
      config.host, config.port, token, n, config.batchSize,
      'compression=false&protocol_version=1'
    );

    if (connections.length === 0) {
      console.log('    FAILED: no connections');
      continue;
    }

    const actual = connections.length;

    // Each connection pings concurrently
    const pingPromises = connections.map((ws) => pingLoop(ws));
    const pingResults = await Promise.allSettled(pingPromises);

    const merged = new LatencyHistogram();
    const wsList: protocol.WsConnection[] = [];

    for (const r of pingResults) {
      if (r.status === 'fulfilled') {
        merged.merge(r.value.hist);
        wsList.push(r.value.ws);
      }
    }

    console.log(`    Total pings: ${merged.len()}`);
    merged.printSummary('    ');

    results.push({
      tier: n,
      connected: actual,
      errors: n - actual,
      durationSecs: 0,
      messagesSent: merged.len(),
      messagesReceived: merged.len(),
      bytesSent: 0,
      latency: merged.isEmpty() ? null : merged.toSummary(),
      extra: {},
    });

    await protocol.closeAll(wsList);
    console.log();
  }

  return results;
}

async function pingLoop(
  ws: protocol.WsConnection
): Promise<{ hist: LatencyHistogram; ws: protocol.WsConnection }> {
  const hist = new LatencyHistogram();

  for (let i = 0; i < PINGS_PER_CONNECTION; i++) {
    const t0 = process.hrtime.bigint();
    const ping = JSON.stringify({
      t: 'ping',
      p: { client_timestamp: Date.now() },
    });

    try {
      await new Promise<void>((resolve, reject) => {
        ws.send(ping, (err) => {
          if (err) reject(err);
          else resolve();
        });
      });
    } catch {
      break;
    }

    // Wait for PONG
    const gotPong = await waitForPong(ws, 5000);
    if (gotPong) {
      const rttUs = Number(process.hrtime.bigint() - t0) / 1000;
      hist.recordUs(Math.max(Math.round(rttUs), 1));
    }
  }

  return { hist, ws };
}

function waitForPong(ws: protocol.WsConnection, timeoutMs: number): Promise<boolean> {
  return new Promise((resolve) => {
    const timer = setTimeout(() => {
      ws.removeListener('message', handler);
      resolve(false);
    }, timeoutMs);

    function handler(data: WebSocket.RawData) {
      const parsed = protocol.parseWseMessage(data);
      if (parsed && protocol.isPong(parsed)) {
        clearTimeout(timer);
        ws.removeListener('message', handler);
        resolve(true);
      }
    }

    ws.on('message', handler);
  });
}
